### PR TITLE
Add support for multiple manifest types

### PR DESCRIPTION
**Pull request by zbiljic**
Builds upon ethersphere/manifest#6

(will be updated after that PR is merged)

Local 'manifest' package (with interfaces) is used, and there external implementations have instance, which can be used in rest of the project. In this case only API package.

Here initial "simple" implementation is used, which mostly replaces existing code (without the need for extra headers and name found in previous code).

Different manifest implementations are differentiated by content type metadata of the manifest file. Here the simple implementation uses `application/bzz-manifest-simple+json` content type (open for comments).

### CODE REVIEW

**Review by janos — COMMENTED** [review]
Very nice!

File: go.mod
Comment (zbiljic) [reply go.mod:11]: This must be changed once that other PR is merged.
In regards to versioning, I am for it, however we have hash anyway.

File: pkg/api/dirs.go
Comment (zbiljic) [reply pkg/api/dirs.go:147]: Copied from some previous code, will change.

**Review by janos — APPROVED** [review]
LGTM 👍

File: pkg/api/bzz.go
Comment (acud) [inline pkg/api/bzz.go:83]: nitpicking but we usually don't have line feeds between parameters on a method call

File: pkg/manifest/manifest.go
Comment (acud) [inline pkg/manifest/manifest.go:42]: so now a manifest entry only knows to represent an address and that's it?

File: pkg/manifest/manifest.go
Comment (acud) [inline pkg/manifest/manifest.go:57]: see comment above (this goes for the entire PR)

File: pkg/manifest/simple.go
Comment (acud) [inline pkg/manifest/simple.go:30]: i am not entirely sure this is desirable to have on the manifest level. you might create a manifest and pass it on to some other calling context to handle and that way you lose control over cancellations in deeper calls in the cascade

**Review by acud — APPROVED** [review]

File: pkg/api/bzz.go
Comment (zbiljic) [reply pkg/api/bzz.go:83]: I opted for this (on several places) because I figured it would be better (easier to read and compare) longer lines.
There is no official requirement, however my thoughts were that once the line goes over 100 characters it might be smart to consider other options (personally I try to go to 80). I also find limited line easier for terminal (but this might be due to my configuration).
If we are to decide to push single-line only, that is fine by me, we can add it to `CODING.md`, and I will change these.

File: pkg/manifest/manifest.go
Comment (zbiljic) [reply pkg/manifest/manifest.go:42]: Yes, that is the change (that I mentioned in the description of this PR).

Name, as you previously noted should be removed as it is a bit pointless in this structure, can just be extracted from the path using `filepath.Base` at any point. As for the headers go, only header that is being added is the `Content-Type` header during upload (again, based on file name using `mime.TypeByExtension`).

Both of these things were also being saved in the actual file on the Swarm in the special `entry.Entry` object, with reference to `entry.Metadata` which contains filename and mime type.

At the moment there is no way (at least I count not find one) to modify the name of already uploaded file after the fact. Also there is no way to add additional headers to the file, or modify existing manifest (through some API) to add additional headers.

Regarding uploading TAR collection using `/dirs` API: Marcelo and myself have been working on separate parts for "website support", and while I added initial manifest implementation (that had 'name' and 'headers') and serving content via `/bzz` API, actual expectation I had was that in that case content would be saved as is using 'bytes' API, while Marcelo used 'files' API (saving files with metadata) on the `/dirs` API side.
During the merge time we had a meeting with Rinke and Viktor regarding this, and decision was to use 'files' for storing files, meaning manifest will contain references to files, which themselves already have metadata which was initially added in manifest. Since it is already located there, I see no point in having it here also.

Now, we could change this back and say that we want to save only bytes for manifest uploaded files, in which case it is easy to return this to previous state.

P.S.
Important note for this is that the initial `mantaray` implementation does not have support for complex content inside. This can easily be changed though.
Also I would like to note that this fact was not the reason for removing fields.

File: pkg/manifest/simple.go
Comment (zbiljic) [reply pkg/manifest/simple.go:30]: Personally I never saw manifest as something that was suppose to be shared in any special way around the code, apart from few simple operations with it. Until now there were not usages of the context in regards with manifest.

As far as the context goes, it is only ever used to save or load manifest.
Load will happen during actual constructor, while the save might be separated, however it could not devise a way to initialize manifest at single point, add stuff from multiple context-es, and save in some completely other one.

It will be easy to change this anyway.

**Comment by Eknir** [comment]
@zbiljic , as far as I understood, this will be for the next release. Is this still the case?

**Comment by zbiljic** [comment]
@Eknir I cannot give answer to that question.

As mentioned in the plan there are several steps that were intended to be completed. This is just one (of several) PR's.
Multiple questions/concerns were raised, and I do not know what are design decisions regarding some parts.
I plan to write architectural proposal explaining our state, and possible directions, so that we have somewhere to discuss, and finalize decisions.

As far as the amount of time needed for each PR, it depends on multiple factors, and that is also one of things that might decide whether this is completed in time.

**Comment by Eknir** [comment]
@zbiljic , thanks for the quick answer. I see this PR already has 2 approvals, so that's already good :). Good suggestion to make architecture proposals, but also let's make sure that the upcoming release contains what we want it to contain. I'm happy to jump on a call on this topic, perhaps with @acud and @janos , to make sure we are putting the effort where it is most effective.
Thanks again.

**Comment by zbiljic** [comment]
@Eknir Related PR from manifest project (noted in the description here) needs to be merged first, after which this to be changed, before this one can be merged.

File: pkg/manifest/manifest.go
Comment (Eknir) [reply pkg/manifest/manifest.go:42]: Related: https://github.com/ethersphere/bee/issues/463#issuecomment-677590074

**Review by acud — APPROVED** [review]

**Pull request merged**
